Docs for synthetic source (#87416)
elastic/elasticsearch: docs/reference/mapping/fields/synthetic-source.asciidoc

[[synthetic-source]]
==== Synthetic `_source`

Though very handy to have around, the source field takes up a significant amount
of space on disk. Instead of storing source documents on disk exactly as you

> send them, Elasticsearch can reconstruct source content on the fly. Enable this

Reviewer: Elasticsearch can reconstruct source content on the fly **upon retrieval**?

** <<ip-synthetic-source,`ip`>>
** <<keyword-synthetic-source,`keyword`>>
** <<numeric-synthetic-source,`long`>>
** <<numeric-synthetic-source,`scaled_float`>>
** <<numeric-synthetic-source,`short`>>
** <<text-synthetic-source,`text`>> (with a `keyword` sub-field)

[[synthetic-source-modifications]]
===== Synthetic source modifications

When synthetic `_source` is enabled, retrieved documents undergo some
modifications compared to the original JSON.

[[synthetic-source-modifications-leaf-arrays]]
====== Arrays moved to leaf fields
Synthetic `_source` arrays are moved to leaves. For example:

[source,console,id=synthetic-source-leaf-arrays-example]
----
PUT idx/_doc/1
{
  "foo": [
    {
      "bar": 1
    },
    {
      "bar": 2
    }
  ]
}
----
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

Will become:

[source,console-result]
----
{
  "foo": {
    "bar": [1, 2]
  }
}
----
// TEST[s/^/{"_source":/ s/\n$/}/]

[[synthetic-source-modifications-field-names]]
====== Fields named as they are mapped
Synthetic source names fields as they are named in the mapping. When used
with <<dynamic,dynamic mapping>>, fields with dots (`.`) in their names are, by

> default, interpreted as multiple objects. For example:

Reviewer: , while dots in field names are preserved within objects that have subobjects disabled (with a link to the docs for the subobjects paramater)

[[synthetic-source]]
==== Synthetic `_source`

> Though very handy to have around, the source field takes up a fair amount of
> space on disk. Instead of storing it on disk precisely as you sent it, Elasticsearch
> can reconstruct it on the fly. Enable this by setting `synthetic: true` in `_source`:

Reviewer 2: ```suggestion
Though very handy to have around, the source field takes up a significant amount
of space on disk. Instead of storing source documents on disk exactly as you
send them, Elasticsearch can reconstruct source content on the fly. Enable this
by setting `synthetic: true` in `_source`:
```

[source,console,id=enable-synthetic-source-example]
----
PUT idx
{
  "mappings": {
    "_source": {
      "synthetic": true
    }
  }
}
----
// TESTSETUP

> This on the fly reconstruction <<synthetic-source-modifications,modifies>> is *generally*
> slower than saving the source precisely and loading it, but it saves a lot of space.
> It also modifies the `_source` and is only supported if the index is entirely made up
> of the following field types:
> 
> * <<boolean-synthetic-source,`boolean`>>
> * <<numeric-synthetic-source,`byte`>>
> * <<numeric-synthetic-source,`double`>>
> * <<numeric-synthetic-source,`float`>>
> * <<geo-point-synthetic-source,`geo_point`>>
> * <<numeric-synthetic-source,`half_float`>>
> * <<numeric-synthetic-source,`integer`>>
> * <<ip-synthetic-source,`ip`>>
> * <<keyword-synthetic-source,`keyword`>>
> * <<numeric-synthetic-source,`long`>>
> * <<numeric-synthetic-source,`scaled_float`>>
> * <<numeric-synthetic-source,`short`>>
> * <<text-synthetic-source,`text`>>

Reviewer 2: ```suggestion
While this on the fly reconstruction is *generally* slower than saving the source
documents verbatim and loading them at query time, it saves a lot of storage
space. There are a couple of other restrictions to be aware of:

* When you retrieve sythetic `_source` content it undergoes minor
<<synthetic-source-modifications,modifications>> compared to the original JSON.
* Synthetic `_source` can be used with indices that contain only these field
types:

** <<boolean-synthetic-source,`boolean`>>
** <<numeric-synthetic-source,`byte`>>
** <<numeric-synthetic-source,`double`>>
** <<numeric-synthetic-source,`float`>>
** <<geo-point-synthetic-source,`geo_point`>>
** <<numeric-synthetic-source,`half_float`>>
** <<numeric-synthetic-source,`integer`>>
** <<ip-synthetic-source,`ip`>>
** <<keyword-synthetic-source,`keyword`>>
** <<numeric-synthetic-source,`long`>>
** <<numeric-synthetic-source,`scaled_float`>>
** <<numeric-synthetic-source,`short`>>
** <<text-synthetic-source,`text`>>
```
I suggest breaking out the restrictions as separate bullets and rephrasing slightly.

[[synthetic-source-modifications]]

> ===== Synthetic source modifications

Reviewer 2: ```suggestion
===== Synthetic source modifications

When synthetic `_source` is enabled, retrieved documents undergo some
modifications compared to the original JSON.
```

[[synthetic-source-modifications-alphabetical]]

> ====== Sorts fields alphabetically
> Synthetic source will make sort all fields alphabetically so:

Reviewer 2: ```suggestion
====== Alphabetical sorting
Synthetic `_source` fields are sorted alphabetically. For example:
```
I think the headings might read a bit better as nouns.

[source,console,id=synthetic-source-sorted-example]
----
PUT idx/_doc/1
{
  "foo": 1,
  "bar": 2,
  "baz": 3
}
----
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

> Will become:

Reviewer 2: ```suggestion
Is retrieved as:
```

[source,console-result]
----
{
  "bar": 2,
  "baz": 3,
  "foo": 1
}
----
// TEST[s/^/{"_source":/ s/\n$/}/]

[[synthetic-source-modifications-leaf-arrays]]

> ====== Moves arrays to leaf fields
> Synthetic source will move all arrays to leaves so:

Reviewer 2: ```suggestion
====== Arrays moved to leaf fields
Synthetic `_source` arrays are moved to leaves. For example:
```

[source,console,id=synthetic-source-leaf-arrays-example]
----
PUT idx/_doc/1
{
  "foo": [
    {
      "bar": 1
    },
    {
      "bar": 2
    }
  ]
}
----
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

Will become:

[source,console-result]
----
{
  "foo": {
    "bar": [1, 2]
  }
}
----
// TEST[s/^/{"_source":/ s/\n$/}/]

[[synthetic-source-modifications-field-names]]

> ====== Names fields as they are named in the mapping
> Synthetic source will name fields as they are mapped. <<dynamic,dynamic mapping>>
> defaults to interpreting fields with dots in their names as objects so dynamic
> mapping will make documents as "objecty" as possible so:

Reviewer 2: ```suggestion
====== Fields named as they are mapped
Synthetic source names fields as they are named in the mapping. When used
with <<dynamic,dynamic mapping>>, fields with dots (`.`) in their names are, by
default, interpreted as multiple objects. For example:
```

[source,console,id=synthetic-source-objecty-example]
----
PUT idx/_doc/1
{
  "foo.bar.baz": 1
}
----
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

Will become:

[source,console-result]
----
{
  "foo": {
    "bar": {
      "baz": 1
    }
  }
}
----
// TEST[s/^/{"_source":/ s/\n$/}/]

> You can keep dots in the names by settings <<subobjects>> to `false` so:

Reviewer 2: ```suggestion
To change this behavior and store names without splitting them on dots, set
<<subobjects>> to `false`. For example:```

[source,console,id=synthetic-dot-example]
----
PUT idx
{
  "mappings": {
    "subobjects": false,
    "_source": {
      "synthetic": true
    }
  }
}

PUT idx/_doc/1
{
  "foo.bar.baz": 1
}
----
// TEST[s/^/DELETE idx\n/]
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

> Will stay:

Reviewer 2: ```suggestion
Will remain as:
```

[[synthetic-source]]
==== Synthetic `_source`

Though very handy to have around, the source field takes up a significant amount
of space on disk. Instead of storing source documents on disk exactly as you
send them, Elasticsearch can reconstruct source content on the fly. Enable this
by setting `synthetic: true` in `_source`:

[source,console,id=enable-synthetic-source-example]
----
PUT idx
{
  "mappings": {
    "_source": {

>       "synthetic": true

Reviewer: not a question for the docs specifically but users may wonder who takes the precedence between enabled and synthetic here if both are specified?

Author: `enabled: false, synthetic: true` is rejected. Or, at least, it will be. I'm sad to make something like that, but it's kind of required for backwards compatibility.

Reviewer: why is that? folks that have disabled source in the past, could potentially switch to synthetic source if they have all fields with doc_values etc.?

Reviewer: or especially if we were to synthetize from stored fields too?

Author: `synthetic` can't be modified either and old indices default to false. Mostly because I didn't want to have to think about what enabling it would mean. I hadn't really thought about folks with large old indices and source disabled. For what it's worth, there is a good chance it won't work either - it is the `picky` implementation of synthetic source after all.

Reviewer: I see, maybe this could be something to consider and think about as a follow-up

    }
  }
}
----
// TESTSETUP

> While this on the fly reconstruction is *generally* slower than saving the source

Reviewer: it is slower at search time but doesn't it make indexing faster? Or is the indexing speed gain negligible?

Author: It's complicated. We still make `_recovery_source` so we still have to write the source, but we remove that `_recovery_source` pretty darn quick. I don't expect really much indexing performance change, but merging should be marginally faster.

documents verbatim and loading them at query time, it saves a lot of storage
space. There are a couple of restrictions to be aware of:

* When you retrieve sythetic `_source` content it undergoes minor
<<synthetic-source-modifications,modifications>> compared to the original JSON.
* Synthetic `_source` can be used with indices that contain only these field
types:

** <<boolean-synthetic-source,`boolean`>>
** <<numeric-synthetic-source,`byte`>>
** <<numeric-synthetic-source,`double`>>
** <<numeric-synthetic-source,`float`>>
** <<geo-point-synthetic-source,`geo_point`>>
** <<numeric-synthetic-source,`half_float`>>
** <<numeric-synthetic-source,`integer`>>
** <<ip-synthetic-source,`ip`>>
** <<keyword-synthetic-source,`keyword`>>
** <<numeric-synthetic-source,`long`>>
** <<numeric-synthetic-source,`scaled_float`>>
** <<numeric-synthetic-source,`short`>>

> ** <<text-synthetic-source,`text`>>

Reviewer: isn't this half the truth given that text has to have a keyword sub-field?

Author: The link says that, yeah. Do you think I should add it in the list?

Reviewer: Got it, maybe add (with a keyword sub-field) in the list?

Author: :+1:

Author: 5962c2716ea

[[synthetic-source-modifications]]
===== Synthetic source modifications

When synthetic `_source` is enabled, retrieved documents undergo some
modifications compared to the original JSON.

[[synthetic-source-modifications-alphabetical]]
====== Alphabetical sorting
Synthetic `_source` fields are sorted alphabetically. For example:

[source,console,id=synthetic-source-sorted-example]
----
PUT idx/_doc/1
{
  "foo": 1,
  "bar": 2,
  "baz": 3
}
----
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

Will become:

[source,console-result]
----
{
  "bar": 2,
  "baz": 3,
  "foo": 1

> }

Reviewer: reading this, as much as I understand why you mention it, keys ordering is never a guarantee in JSON land. I wonder if it's then needed to provide an example. Could we shorten it and say "You should never rely on keys ordering, but if you do beware you'll lose that with synthetic source as what you'll get back is not 100% what you sent"?

Author: I get that. I'm providing an example for everything so I like the.... balance? I don't know the right word. I could link to the spec and mention that ordering isn't supported by the spec.

Reviewer: I don't have a strong opinion but while the examples with arrays ought to be mentioned, taking this one out would make the docs page a little more compact without losing much?

Author: We need a tie breaker. @romseygeek, break a tie!

Reviewer: No pressure :)

Reviewer 3: I'm with @javanna, we don't guarantee key ordering anywhere so I don't think an example is necessary here.

Author: Democracy wins. I'll remove the example. I'll reduce this to a note.

Will become:

[source,console-result]
----
{
  "foo": {
    "bar": [1, 2]
  }
}
----
// TEST[s/^/{"_source":/ s/\n$/}/]

[[synthetic-source-modifications-field-names]]
====== Fields named as they are mapped
Synthetic source names fields as they are named in the mapping. When used
with <<dynamic,dynamic mapping>>, fields with dots (`.`) in their names are, by
default, interpreted as multiple objects. For example:

[source,console,id=synthetic-source-objecty-example]
----
PUT idx/_doc/1
{
  "foo.bar.baz": 1
}
----
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

Will become:

[source,console-result]
----
{
  "foo": {
    "bar": {
      "baz": 1
    }
  }
}
----
// TEST[s/^/{"_source":/ s/\n$/}/]

To change this behavior and store names without splitting them on dots, set
<<subobjects>> to `false`. For example:```

[source,console,id=synthetic-dot-example]
----
PUT idx
{
  "mappings": {
    "subobjects": false,
    "_source": {
      "synthetic": true
    }
  }
}

PUT idx/_doc/1
{
  "foo.bar.baz": 1
}
----
// TEST[s/^/DELETE idx\n/]
// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]

Will remain as:

[source,console-result]
----
{
  "foo.bar.baz": 1
}

> ----

Reviewer: I wonder what is specific of synthetic source about how fields are mapped: isn't this all about dynamic mappings? Did you mean to say that synthetic _source recreate the objects structure or not depending on how fields are mapped?

Author: It recreates the structure precisely as the objects are mapped. At first I tried to explicitly create a field mapped as `foo.bar.baz` but the mapping infrastructure unraveled it so I went with this. I'll make an example, one moment.

Reviewer: Ok, I guess I was wondering why this needs to be specifically explained, isn't it the expected behaviour? :)

Reviewer: I think you are explaining how dynamic mappings work in combination with subobjects fails, and that synthetic source respects that, but you could omit this here, because it would be surprising and would need to be mentioned if it worked otherwise.

Author: I suppose. But it was kind of surprising to me that it *did* work with subobjects. Maybe that's just stuck in my head.

Reviewer: Scratch that, maybe this is not so obvious :)
One thing to mention could be that you always recreate the object structure despite dots in fields names were provided in the first place. Why not always flatten by the way? After all, there are two ways to provide that document and that leads to the same mapping, you have 50% chances to pick the variant that was sent in the first place.

By the way we may start accepting docs with subobjects even when subobjects are disabled and treat them like dotted fields, but I think the current behaviour of "flattening" would still be good in that case.

Author: I've removed the section

Author: Oh no! I removed the section!

> Why not always flatten by the way?

We asked some kibana friends and they said folks liked things shaped this way instead of flattened. The objects feel right to folks. Also, I figured if we were guessing anyway the "more objecty" approach was probably more likely to match what users sent. I figured flattened fields with dots in them is more rare.

Reviewer: fine with me. I think you removed only the section with subobjects false and you could make the remaining section about recreating objects shorter by saying that you follow the mappings structure hence prefer nested objects over dots in fields names but the opposite when subobjects are disabled.

Author: > I think you removed only the section with subobjects false and you could make the remaining section about recreating objects shorter by saying that you follow the mappings structure hence prefer nested objects over dots in fields names but the opposite when subobjects are disabled.

Two things:
1. Don't use the word `nested` unless you mean it. I get scared.
2. Do you want me to change the words above? Could you suggest a change? I think the "dynamic mappings work like BLAH" bit is useful because it explains where those objects come from. I don't mean to explain all of dynamic mappings, just enough to give folks a hint.
